File: frontend/src/api/quality.ts
import client from './client'

export interface QCDefect {
  id: number
  loai_loi: string
  mo_ta: string | null
  so_luong_loi: number
  hinh_anh_path: string | null
}

export interface QCSheet {
  id: number
  so_phieu: string
  loai: string
  ref_type: string | null
  ref_id: number | null
  ngay: string
  nguoi_kiem_tra: string | null
  ket_qua: string | null
  ghi_chu: string | null
  phap_nhan_id: number | null
  phan_xuong_id: number | null
  created_by: number | null
  created_at: string
  defects: QCDefect[]
}

export interface QCStats {
  tong: number
  dat: number
  khong_dat: number
  tam_chap_nhan: number
  chua_co_ket_qua: number
  ty_le_dat_pct: number
}

export interface CreateQCSheetPayload {
  loai: string
  ref_type?: string | null
  ref_id?: number | null
  ngay: string
  nguoi_kiem_tra?: string | null
  ket_qua?: string | null
  ghi_chu?: string | null
  phap_nhan_id?: number | null
  phan_xuong_id?: number | null
  defects?: Array<{ loai_loi: string; mo_ta?: string; so_luong_loi?: number }>
}

export interface UpdateKetQuaPayload {
  ket_qua?: string | null
  nguoi_kiem_tra?: string | null
  ghi_chu?: string | null
  defects?: Array<{ loai_loi: string; mo_ta?: string; so_luong_loi?: number }>
}

export const qualityApi = {
  list: (params?: Record<string, string | number | null | undefined>) =>
    client.get<QCSheet[]>('/api/qc-sheets', { params }).then(r => r.data),

  get: (id: number) =>
    client.get<QCSheet>(`/api/qc-sheets/${id}`).then(r => r.data),

  create: (payload: CreateQCSheetPayload) =>
    client.post<QCSheet>('/api/qc-sheets', payload).then(r => r.data),

  updateKetQua: (id: number, payload: UpdateKetQuaPayload) =>
    client.patch<QCSheet>(`/api/qc-sheets/${id}/ket-qua`, payload).then(r => r.data),

  delete: (id: number) =>
    client.delete(`/api/qc-sheets/${id}`),

  stats: (params?: Record<string, string | null | undefined>) =>
    client.get<QCStats>('/api/qc-sheets/stats', { params }).then(r => r.data),
}

File: backend/app/models/quality.py
from datetime import date, datetime, timezone
from sqlalchemy import Date, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class QCSheet(Base):
    __tablename__ = "qc_sheets"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    so_phieu: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    loai: Mapped[str] = mapped_column(String(20), nullable=False)
    # nhan_hang | san_xuat | xuat_hang

    ref_type: Mapped[str | None] = mapped_column(String(50))
    # goods_receipt | production_order | delivery_order
    ref_id: Mapped[int | None] = mapped_column(Integer)

    ngay: Mapped[date] = mapped_column(Date, nullable=False)
    nguoi_kiem_tra: Mapped[str | None] = mapped_column(String(100))
    ket_qua: Mapped[str | None] = mapped_column(String(20))
    # dat | khong_dat | tam_chap_nhan
    ghi_chu: Mapped[str | None] = mapped_column(Text)

    phap_nhan_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phap_nhan.id"))
    phan_xuong_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phan_xuong.id"))
    created_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(
            timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc))

    phap_nhan = relationship("PhapNhan")
    phan_xuong = relationship("PhanXuong")
    defects: Mapped[list["QCDefect"]] = relationship("QCDefect", back_populates="sheet", cascade="all, delete-orphan")


class QCDefect(Base):
    __tablename__ = "qc_defects"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    qc_sheet_id: Mapped[int] = mapped_column(Integer, ForeignKey("qc_sheets.id", ondelete="CASCADE"), nullable=False)
    loai_loi: Mapped[str] = mapped_column(String(100), nullable=False)
    mo_ta: Mapped[str | None] = mapped_column(Text)
    so_luong_loi: Mapped[int] = mapped_column(Integer, default=0)
    hinh_anh_path: Mapped[str | None] = mapped_column(String(500))

    sheet: Mapped["QCSheet"] = relationship("QCSheet", back_populates="defects")


class QCGiayCuonPhieu(Base):
    __tablename__ = "qc_giay_cuon_phieu"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    so_phieu: Mapped[str] = mapped_column(String(40), unique=True, nullable=False)

    # Liên kết
    paper_material_id: Mapped[int] = mapped_column(Integer, ForeignKey("paper_materials.id"), nullable=False)
    goods_receipt_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("goods_receipts.id"))
    goods_receipt_item_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("goods_receipt_items.id"))

    # Thông tin phiếu
    ngay_nhap_giay: Mapped[date | None] = mapped_column(Date)
    ngay_kiem_tra: Mapped[date] = mapped_column(Date, nullable=False)
    nguoi_kiem_tra: Mapped[str | None] = mapped_column(String(100))
    trong_luong_tem: Mapped[float | None] = mapped_column(Float)   # KG trên nhãn
    kho_thuc_te: Mapped[float | None] = mapped_column(Float)       # cm — khổ đo thực tế
    kho_tc: Mapped[float | None] = mapped_column(Float)            # cm — khổ tiêu chuẩn (snapshot)

    # Snapshot tiêu chuẩn tại thời điểm kiểm tra (audit trail)
    tc_dinh_luong: Mapped[float | None] = mapped_column(Float)
    tc_sai_so_pct: Mapped[float | None] = mapped_column(Float)
    tc_do_buc: Mapped[float | None] = mapped_column(Float)
    tc_do_nen_vong: Mapped[float | None] = mapped_column(Float)

    # Định lượng GSM
    dl_l1: Mapped[float | None] = mapped_column(Float)
    dl_l2: Mapped[float | None] = mapped_column(Float)
    dl_tb: Mapped[float | None] = mapped_column(Float)
    dl_ket_qua: Mapped[str | None] = mapped_column(String(20))     # dat | khong_dat

    # Độ bục (kgf/cm²)
    buc_l1: Mapped[float | None] = mapped_column(Float)
    buc_l2: Mapped[float | None] = mapped_column(Float)
    buc_l3: Mapped[float | None] = mapped_column(Float)
    buc_l4: Mapped[float | None] = mapped_column(Float)
    buc_tb: Mapped[float | None] = mapped_column(Float)
    buc_ket_qua: Mapped[str | None] = mapped_column(String(20))

    # Độ nén vòng (kgf/6inch)
    nen_vong_l1: Mapped[float | None] = mapped_column(Float)
    nen_vong_l2: Mapped[float | None] = mapped_column(Float)
    nen_vong_l3: Mapped[float | None] = mapped_column(Float)
    nen_vong_tb: Mapped[float | None] = mapped_column(Float)
    nen_vong_ket_qua: Mapped[str | None] = mapped_column(String(20))

    # Khổ giấy pass/fail: (tc-4) ≤ thực tế ≤ (tc+4)
    kho_ket_qua: Mapped[str | None] = mapped_column(String(20))

    # Kết quả tổng
    ket_qua: Mapped[str | None] = mapped_column(String(20))        # dat | khong_dat
    ghi_chu: Mapped[str | None] = mapped_column(Text)

    phap_nhan_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("phap_nhan.id"))
    created_by: Mapped[int | None] = mapped_column(Integer, ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc)
    )

    paper_material = relationship("PaperMaterial", foreign_keys=[paper_material_id])
    goods_receipt = relationship("GoodsReceipt", foreign_keys=[goods_receipt_id])
    phap_nhan = relationship("PhapNhan")
